core/modules/qos.cc

```diff
@@ -0,0 +1,441 @@
+    }
+
+    size_acc += f.size;
+  }
+  default_gate_ = LOOKUP_FAIL_GATE;
+  total_key_size_ = align_ceil(size_acc, sizeof(uint64_t));
+  for (int i = 0; i < arg.values_size(); i++) {
+    const auto &field = arg.values(i);
+    CommandResponse err;
+    values_.emplace_back();
+    struct MeteringField &f = values_.back();
+    f.pos = value_acc;
+    err = AddFieldOne(field, &f, ValueType);
+    if (err.error().code() != 0) {
+      return err;
+    }
+
+    value_acc += f.size;
+  }
+
+  total_value_size_ = align_ceil(value_acc, sizeof(uint64_t));
+
+  uint8_t *cs = (uint8_t *)&mask;
+  for (int i = 0; i < size_acc; i++) {
+    cs[i] = 0xff;
+  }
+
+  table_.Init(total_key_size_);
+  return CommandSuccess();
+}
+
+void Qos::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
+  gate_idx_t default_gate;
+  MeteringKey keys[bess::PacketBatch::kMaxBurst] __ymm_aligned;
+  bess::Packet *pkt = nullptr;
+  default_gate = ACCESS_ONCE(default_gate_);
+
+  int cnt = batch->cnt();
+  struct value *val[cnt];
+  default_gate = ACCESS_ONCE(default_gate_);
+
+  for (const auto &field : fields_) {
+    int offset;
+    int pos = field.pos;
+    int attr_id = field.attr_id;
+
+    if (attr_id < 0) {
+      offset = field.offset;
+    } else {
+      offset = bess::Packet::mt_offset_to_databuf_offset(attr_offset(attr_id));
+    }
+
+    for (int j = 0; j < cnt; j++) {
+      char *buf_addr = batch->pkts()[j]->buffer<char *>();
+
+      /* for offset-based attrs we use relative offset */
+      if (attr_id < 0) {
+        buf_addr += batch->pkts()[j]->data_off();
+      }
+
+      char *key = reinterpret_cast<char *>(keys[j].u64_arr) + pos;
+
+      *(reinterpret_cast<uint64_t *>(key)) =
+          *(reinterpret_cast<uint64_t *>(buf_addr + offset));
+
+      size_t len = reinterpret_cast<size_t>(total_key_size_ / sizeof(uint64_t));
+
+      for (size_t i = 0; i < len; i++) {
+        keys[j].u64_arr[i] = keys[j].u64_arr[i] & mask[i];
+      }
+    }
+  }
+
+  uint64_t hit_mask = table_.Find(keys, val, cnt);
+
+  for (int init = 0; init < cnt; init++) {
+    if ((hit_mask & (1ULL << init)) == 0) {
+      EmitPacket(ctx, batch->pkts()[init], default_gate);
+      // std::cout << "qer default gate " << default_gate << std::endl;
+      continue;
+    }
+
+    /* if lookup was successful, then set values (if possible) */
+    if (hit_mask & (1 << init)) {
+      // std::cout << "qer gate " << val[init]->ogate << std::endl;
+      if ((val[init]->ogate == CONFIG_DROP_GATE) ||
+          (val[init]->ogate == METER_GREEN_GATE) ||
+          (val[init]->ogate == METER_YELLOW_GATE) ||
+          (val[init]->ogate == METER_RED_GATE)) {
+        // std::cout << "qer drop " << val[init]->ogate << std::endl;
+        EmitPacket(ctx, batch->pkts()[init], CONFIG_DROP_GATE);
+      } else if (val[init]->ogate == UNMETERED_GATE) {
+        // std::cout << "qer unmeter " << val[init]->ogate << std::endl;
+        EmitPacket(ctx, batch->pkts()[init], UNMETERED_GATE);
```

Comment:
We can remove this by using a single `EmitPacket(ctx, batch->pkts()[init], ogate); ` in line 258

Reply:
The whole if block would look like this 

```cpp
    if (hit_mask & (1 << init)) {
        if (val[init]->ogate == METER_GATE) {
        // meter

        // update ogate to color-specific gate

        }

        // update values
        EmitPacket(ctx, batch->pkts()[init], ogate);
    }
```